homework_3/src/main/java/com/epam/spring/service/impl/ProductServiceImpl.java

```diff
@@ -0,0 +1,92 @@
+package com.epam.spring.service.impl;
+
+import com.epam.spring.controller.dto.ProductDto;
+import com.epam.spring.service.ProductService;
+import com.epam.spring.service.mapper.ProductMapper;
+import com.epam.spring.service.repository.ProductRepository;
+import lombok.AllArgsConstructor;
+import org.springframework.stereotype.Service;
+
+import java.util.List;
+import java.util.stream.Collectors;
+
+@Service
+@AllArgsConstructor
+public class ProductServiceImpl implements ProductService {
+
+    private final ProductRepository productRepository;
+    private final ProductMapper productMapper;
+
+    @Override
+    public ProductDto getProduct(String name) {
+        return productMapper.mapProductDto(productRepository.getProduct(name));
+    }
+
+    @Override
+    public ProductDto getProduct(int id) {
+        return productMapper.mapProductDto(productRepository.getProduct(id));
+    }
+
+    @Override
+    public ProductDto getProduct(String name, String language) {
+        ProductDto product = getProduct(name);
+        product.getProductDescriptionList().removeIf(p -> !p.getLanguage().getShortName().equals(language));
```

Comment:
Hi Viktor. Lambda expression `p -> !p.getLanguage().getShortName().equals(language)` is used several times. To eliminate duplication we can try to place this logic once on the class level:
```
private static final Function<String, Predicate<ProductDescriptionDto>> isNotLanguagePredicate = l -> p -> !p.getLanguage()
            .getShortName()
            .equals(l);
```
Thank you